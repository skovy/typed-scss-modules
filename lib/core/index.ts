export { alerts, setAlertsLogLevel } from "./alerts";
export { generate } from "./generate";
export { listDifferent } from "./list-different";
export { CLIOptions, ConfigOptions } from "./types";
export { watch } from "./watch";
export { writeFile } from "./write-file";
