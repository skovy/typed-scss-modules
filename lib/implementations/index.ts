import nodeSass from "node-sass";
import sass from "node-sass";

/**
 * A list of all possible SASS package implementations that can be used to
 * perform the compilation and parsing of the SASS files. The expectation is
 * that they provide a nearly identical API so they can be swapped out but
 * all of the same logic can be reused.
 */
export const IMPLEMENTATIONS = ["node-sass", "sass"] as const;
export type Implementations = typeof IMPLEMENTATIONS[number];

type Implementation = typeof nodeSass | typeof sass;

/**
 * Determine which default implementation to use by checking which packages
 * are actually installed and available to use.
 *
 * @param resolver DO NOT USE - this is unfortunately necessary only for testing.
 */
export const getDefaultImplementation = (
  resolver: RequireResolve = require.resolve
): Implementations => {
  let pkg: Implementations = "node-sass";

  try {
    resolver("node-sass");
  } catch (error) {
    try {
      resolver("sass");
      pkg = "sass";
    } catch (ignoreError) {
      pkg = "node-sass";
    }
  }

  return pkg;
};

/**
 * Retrieve the desired implementation.
 *
 * @param implementation the desired implementation.
 */
export const getImplementation = (
  implementation?: Implementations
): Implementation => {
  if (implementation === "sass") {
    return require("sass");
  } else {
    return require("node-sass");
  }
};
