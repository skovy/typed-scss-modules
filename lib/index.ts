export { main as default } from "./main";
