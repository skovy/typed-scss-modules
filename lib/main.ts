import fs from "fs";
import path from "path";
import slash from "slash";

import {
  watch,
  MainOptions,
  generate,
  listDifferent,
  setAlertsLogLevel,
} from "./core";

export const main = async (pattern: string, options: MainOptions) => {
  setAlertsLogLevel(options.logLevel);

  // When the provided pattern is a directory construct the proper glob to find
  // all .scss files within that directory. Also, add the directory to the
  // included paths so any imported with a path relative to the root of the
  // project still works as expected without adding many include paths.
  if (fs.existsSync(pattern) && fs.lstatSync(pattern).isDirectory()) {
    if (Array.isArray(options.includePaths)) {
      options.includePaths.push(pattern);
    } else {
      options.includePaths = [pattern];
    }

    // When the pattern provide is a directory, assume all .scss files within.
    pattern = slash(path.resolve(pattern, "**/*.scss"));
  }

  if (options.listDifferent) {
    listDifferent(pattern, options);
    return;
  }

  if (options.watch) {
    watch(pattern, options);
  } else {
    await generate(pattern, options);
  }
};
