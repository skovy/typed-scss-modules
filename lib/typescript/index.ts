export {
  bannerTypeDefault,
  classNamesToTypeDefinitions,
  exportTypeDefault,
  exportTypeInterfaceDefault,
  exportTypeNameDefault,
  quoteTypeDefault,
  ExportType,
  QuoteType,
  EXPORT_TYPES,
  QUOTE_TYPES,
} from "./class-names-to-type-definition";
export { logLevelDefault, LogLevel, LOG_LEVELS } from "../core/alerts";
export { getTypeDefinitionPath } from "./get-type-definition-path";
