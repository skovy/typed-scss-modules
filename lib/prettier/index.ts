import { alerts } from "../core";
import { canResolvePrettier } from "./can-resolve";

import { format, resolveConfig } from "prettier";

interface Prettier {
  format: typeof format;
  resolveConfig: typeof resolveConfig;
}

const isPrettier = (t: unknown): t is Prettier =>
  t &&
  typeof t === "object" &&
  t !== null &&
  "format" in t &&
  typeof (t as Prettier).format === "function" &&
  "resolveConfig" in t &&
  typeof (t as Prettier).resolveConfig === "function";

/**
 * Try to load prettier and config from project to format input,
 * fall back to input if prettier is not found or failed
 *
 * @param {string} input
 */
export const attemptPrettier = async (input: string) => {
  if (!canResolvePrettier()) {
    return input;
  }

  const prettier = require("prettier");
  if (!isPrettier(prettier)) {
    // doesn't look like prettier
    return input;
  }

  try {
    const config = await prettier.resolveConfig(process.cwd(), {
      editorconfig: true,
    });
    // try to return formatted output
    return prettier.format(input, { ...config, parser: "typescript" });
  } catch (error) {
    alerts.notice(`Tried using prettier, but failed with error: ${error}`);
  }

  // failed to format
  return input;
};
