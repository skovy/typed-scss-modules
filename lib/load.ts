import path from "path";
import JoyCon from "joycon";
import { bundleRequire } from "bundle-require";

import { alerts, CLIOptions, ConfigOptions } from "./core";
import { nameFormatDefault } from "./sass";
import {
  bannerTypeDefault,
  exportTypeDefault,
  exportTypeInterfaceDefault,
  exportTypeNameDefault,
  quoteTypeDefault,
  logLevelDefault,
} from "./typescript";
import { getDefaultImplementation } from "./implementations";

const CURRENT_WORKING_DIRECTORY = process.cwd();
const VALID_CONFIG_FILES = [
  "typed-scss-modules.config.ts",
  "typed-scss-modules.config.js",
];
const joycon = new JoyCon();

/**
 * Load a custom config file in the project root directory with any options for this package.
 *
 * This support config files with the format:
 *  - Named `config` export: `export const config = {}`
 *  - Default export: `export default {}`
 */
export const loadConfig = async (): Promise<{} | ConfigOptions> => {
  const configPath = await joycon.resolve(
    VALID_CONFIG_FILES,
    CURRENT_WORKING_DIRECTORY,
    path.parse(CURRENT_WORKING_DIRECTORY).root
  );

  if (configPath) {
    try {
      const configModule = await bundleRequire({
        filepath: configPath,
      });

      const config: ConfigOptions =
        configModule.mod.config || configModule.mod.default || configModule.mod;

      return config;
    } catch (error) {
      alerts.error(
        `An error occurred loading the config file "${configPath}":\n${error}`
      );

      return {};
    }
  }

  return {};
};

// Default values for all options that need defaults.
const DEFAULT_OPTIONS: CLIOptions = {
  nameFormat: nameFormatDefault,
  implementation: getDefaultImplementation(),
  exportType: exportTypeDefault,
  exportTypeName: exportTypeNameDefault,
  exportTypeInterface: exportTypeInterfaceDefault,
  watch: false,
  ignoreInitial: false,
  listDifferent: false,
  ignore: [],
  quoteType: quoteTypeDefault,
  updateStaleOnly: false,
  logLevel: logLevelDefault,
  banner: bannerTypeDefault,
};

/**
 * Given both the CLI and config file options merge into a single options object.
 *
 * When possible, CLI options will override config file options.
 *
 * Some options are only available in the config file. For example, a custom function can't
 * be easily defined via the CLI so some complex options are only available in the config file.
 */
export const mergeOptions = (
  cliOptions: Partial<CLIOptions>,
  configOptions: Partial<ConfigOptions>
): ConfigOptions => {
  return {
    ...DEFAULT_OPTIONS,
    ...configOptions,
    ...cliOptions,
  };
};
