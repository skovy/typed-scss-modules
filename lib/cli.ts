#!/usr/bin/env node

import yargs from "yargs";

import { Aliases, NAME_FORMATS } from "./sass";
import { EXPORT_TYPES, QUOTE_TYPES, LOG_LEVELS } from "./typescript";
import { main } from "./main";
import { IMPLEMENTATIONS } from "./implementations";

const { _: patterns, ...rest } = yargs
  .usage(
    "Generate .scss.d.ts from CSS module .scss files.\nUsage: $0 <glob pattern> [options]"
  )
  .example("$0 src", "All .scss files at any level in the src directory")
  .example(
    "$0 src/**/*.scss",
    "All .scss files at any level in the src directory"
  )
  .example(
    "$0 src/**/*.scss --watch",
    "Watch all .scss files at any level in the src directory that are added or changed"
  )
  .example(
    "$0 src/**/*.scss --includePaths src/core src/variables",
    'Search the "core" and "variables" directory when resolving imports'
  )
  .example(
    "$0 src/**/*.scss --aliases.~name variables",
    'Replace all imports for "~name" with "variables"'
  )
  .example(
    "$0 src/**/*.scss --aliasPrefixes.~ ./node_modules/",
    'Replace the "~" prefix with "./node_modules/" for all imports beginning with "~"'
  )
  .example(
    "$0 src/**/*.scss --ignore **/secret.scss",
    'Ignore any file names "secret.scss"'
  )
  .example(
    "$0 src/**/*.scss --implementation sass",
    "Use the Dart SASS package"
  )
  .example(
    "$0 src/**/*.scss -e default --quoteType double",
    "Use double quotes around class name definitions rather than single quotes."
  )
  .example("$0 src/**/*.scss --logLevel error", "Output only errors")
  .demandCommand(1)
  .option("additionalData", {
    string: true,
    alias: "d",
    describe: "Prepends the SCSS code before each file.",
  })
  .option("aliases", {
    coerce: (obj): Aliases => obj,
    alias: "a",
    describe: "Alias any import to any other value.",
  })
  .option("aliasPrefixes", {
    coerce: (obj): Aliases => obj,
    alias: "p",
    describe: "A prefix for any import to rewrite to another value.",
  })
  .option("nameFormat", {
    choices: NAME_FORMATS,
    alias: "n",
    describe: "The name format that should be used to transform class names.",
  })
  .option("implementation", {
    choices: IMPLEMENTATIONS,
    describe:
      "The SASS package to used to compile. This will default to the sass implementation you have installed.",
  })
  .option("exportType", {
    choices: EXPORT_TYPES,
    alias: "e",
    describe: "The type of export used for defining the type definitions.",
  })
  .option("exportTypeName", {
    string: true,
    describe:
      'Set a custom type name for styles when --exportType is "default."',
  })
  .option("exportTypeInterface", {
    string: true,
    describe:
      'Set a custom interface name for styles when --exportType is "default."',
  })
  .option("watch", {
    boolean: true,
    alias: "w",
    describe:
      "Watch for added or changed files and (re-)generate the type definitions.",
  })
  .option("ignoreInitial", {
    boolean: true,
    describe: "Skips the initial build when passing the watch flag.",
  })
  .option("listDifferent", {
    boolean: true,
    alias: "l",
    describe:
      "List any type definitions that are different than those that would be generated.",
  })
  .option("includePaths", {
    array: true,
    string: true,
    alias: "i",
    describe: "Additional paths to include when trying to resolve imports.",
  })
  .option("ignore", {
    string: true,
    array: true,
    describe: "Add a pattern or an array of glob patterns to exclude matches.",
  })
  .option("outputFolder", {
    string: true,
    alias: "o",
    describe:
      "Define a (relative) folder to output the generated type definitions. Note this requires adding the output folder to tsconfig.json `rootDirs`.",
  })
  .options("quoteType", {
    choices: QUOTE_TYPES,
    alias: "q",
    describe:
      "Specify the quote type so that generated files adhere to your TypeScript rules.",
  })
  .options("updateStaleOnly", {
    boolean: true,
    alias: "u",
    describe:
      "Overwrite generated files only if the source file has more recent changes.",
  })
  .option("logLevel", {
    choices: LOG_LEVELS,
    alias: "L",
    describe: "Verbosity level of console output",
  })
  .options("banner", {
    string: true,
    describe:
      "Inserts text at the top of every output file for documentation purposes.",
  })
  .parseSync();

main(patterns[0] as string, { ...rest });
