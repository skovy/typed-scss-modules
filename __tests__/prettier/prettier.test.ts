import { attemptPrettier } from "../../lib/prettier";
import { classNamesToTypeDefinitions } from "../../lib/typescript";

import prettier from "prettier";

const input =
  "export type Styles = {'myClass': string;'yourClass': string;}; export type Classes = keyof Styles; declare const styles: Styles; export default styles;";

describe("attemptPrettier", () => {
  it("should locate and apply prettier.format", async () => {
    const output = await attemptPrettier(input);
    expect(prettier.format(input, { parser: "typescript" })).toMatch(output);
  });

  it("should match snapshot", async () => {
    const typeDefinition = await classNamesToTypeDefinitions({
      banner: "",
      classNames: ["nestedAnother", "nestedClass", "someStyles"],
      exportType: "default",
    });

    if (!typeDefinition) {
      throw new Error("failed to collect typeDefinition");
    }

    const output = await attemptPrettier(typeDefinition);
    expect(output).toMatchSnapshot();
  });
});

describe("attemptPrettier - mock prettier", () => {
  beforeAll(() => {
    jest.mock("prettier", () => ({
      format: undefined,
    }));
  });

  it("should fail to recognize prettier and return input", async () => {
    const output = await attemptPrettier(input);
    expect(input).toMatch(output);
  });
});

describe("attemptPrettier - mock resolution check", () => {
  beforeAll(() => {
    jest.mock("../../lib/prettier/can-resolve");
  });

  it("should fail to resolve prettier and return input", async () => {
    const output = await attemptPrettier(input);
    expect(input).toMatch(output);
  });
});
