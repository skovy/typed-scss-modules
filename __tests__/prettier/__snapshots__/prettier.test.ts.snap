// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attemptPrettier should match snapshot 1`] = `
"export type Styles = {
  nestedAnother: string;
  nestedClass: string;
  someStyles: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
"
`;
